src/core/mod.rs

```diff
@@ -1,71 +1,79 @@
+use std::convert::TryInto;
+use std::fmt::{Debug, Formatter};
+
 use rand::RngCore;
-use serde::{Deserialize, Serialize};
+use serde::ser::Serialize;
 use sha3::{Digest, Sha3_256};
-use std::convert::TryInto;
 
-#[derive(Serialize, Deserialize, PartialEq, Debug, Clone)]
-pub struct U256(pub [u8; 32]);
+use crate::core::digest::Digests;
+use crate::core::number::U256;
 
-impl U256 {
-    pub fn empty() -> Self {
-        Self([0u8; 32])
-    }
-    pub fn from_bytes(bytes: &[u8]) -> Self {
-        let mut data = [0u8; 32];
-        data[..bytes.len()].copy_from_slice(bytes);
-        Self(data)
-    }
-    pub fn random<R: RngCore>(rng: &mut R) -> Self {
-        let mut data = [0u8; 32];
-        rng.fill_bytes(&mut data);
-        Self(data)
-    }
-    pub fn to_bits(&self) -> [bool; 256] {
-        let mut ret = [false; 256];
-        for i in 0..256 {
-            ret[i] = ((self.0[i / 8] >> (i % 8)) & 1) == 1;
-        }
-        ret
-    }
-    pub fn to_bytes(&self) -> &[u8; 32] {
-        &self.0
-    }
+pub mod blocks;
+pub mod digest;
+pub mod hash;
+pub mod header;
+pub mod number;
+
+macro_rules! auto_trait {
+    (
+        $(
+            $(#[$doc:meta])+
+            trait $name:ident: $( $bound:path ),+;
+        )+
+    ) => {
+        $(
+            $(#[$doc])+
+            pub trait $name: $( $bound + )+ {}
+            impl <T: $($bound +)+> $name for T {}
+        )+
+    };
 }
 
-pub type Signature = u8;
-pub type Hash = U256;
-pub type Money = u32;
+#[macro_export]
+auto_trait!(
+    /// A type that implements Serialize in node runtime
+    trait AutoSerialize: serde::ser::Serialize;
 
-impl Hash {
-    pub fn generate(data: &Vec<u8>) -> Self {
-        let mut hasher = Sha3_256::new();
-        hasher.update(data);
-        Self(hasher.finalize().try_into().unwrap())
-    }
+    /// A type that implements Deserialize in node runtime
+    trait AutoDeserialize: serde::de::DeserializeOwned, serde::ser::Serialize;
+    /// A type that implements Hash in node runtime
+    trait AutoHash: core::hash::Hash;
+    /// A type that implements Display in runtime
+    trait AutoDisplay: core::fmt::Display;
+    /// A type that implements CanBe32Bits
+    trait CanBe32Bits: core::convert::From<u32>;
+);
 
-    // Dummy implementation of a 512-bit hash function, used for generating
-    // scalar and randomness of EdDSA signatures.
-    pub fn generate_two(data: &Vec<u8>) -> (Self, Self) {
-        let mut hasher = Sha3_256::new();
-        hasher.update(data);
-        let first: [u8; 32] = hasher.finalize().try_into().unwrap();
+/// A type that can be used at runtime
+pub trait Field: Send + Sync + Sized + Debug + Clone + Eq + PartialEq + 'static {}
```

Comment:
`Field` could be confused with number fields

Reply:
`Field` maybe a little Puzzling. How about `MemberBound` ? That means if there has a struct in runtime and it's fields should satisfy with the bound.